### PR TITLE
app: wire tracker

### PR DESCRIPTION
**Pull request by corverroos**
Wire tracker in core workflow. Implement failed duty reporter.

Also fix concurrency deadlock between Tracker and Deadliner.

category: feature
ticket: #768

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -44,7 +44,7 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-go-
       - run: docker pull consensys/teku:latest
-      - run: go test -timeout=5m github.com/obolnetwork/charon/app -integration -slow
+      - run: go test -timeout=5m -v github.com/obolnetwork/charon/app -integration -slow
```
Comment (xenowits) [inline .github/workflows/test.yml:47]: discussion: any reason why we need verbose logs?

File: app/app.go [hunk 2]
```diff
@@ -321,6 +322,11 @@ func wireCoreWorkflow(ctx context.Context, life *lifecycle.Manager, conf Config,
 		pubshares = append(pubshares, eth2Share)
 	}
 
+	peers, err := lock.Peers()
```
Comment (xenowits) [inline app/app.go:325]: i think `peers` is needed as a param for `wireTracker` so why not colocate them?

File: core/tracker/tracker.go [hunk 3]
```diff
@@ -67,11 +67,11 @@ type Tracker struct {
 	deadliner core.Deadliner
 	quit      chan struct{}
 
-	// failedDutyReporter instruments the duty. It ignores non-failed duties.
-	failedDutyReporter func(core.Duty, bool, string, string)
+	// failedDutyReporter instruments duty failures.
```
Comment (xenowits) [inline core/tracker/tracker.go:70]: discussion: what's the preferred approach when declaring function literals? omitting the names of the parameters or explicitly naming the parameters?

File: core/tracker/tracker.go [hunk 4]
```diff
@@ -146,9 +150,18 @@ func analyseDutyFailed(duty core.Duty, es []event) (bool, component, string) {
 	return true, events[0].component + 1, fmt.Sprintf("%s failed in %s component", duty.String(), (events[0].component + 1).String())
 }
 
-// failedDutyReporter instruments the duty. It ignores non-failed duties.
-// TODO(xenowits): Implement logic for reporting duties.
-func failedDutyReporter(core.Duty, bool, string, string) {}
+// failedDutyReporter instruments failed duties.
```
Comment (xenowits) [inline core/tracker/tracker.go:153]: logs and instruments failed duties?

**Review by xenowits — APPROVED** [review]
Looks good!

File: .github/workflows/test.yml (hunk 1 above)
Comment (corverroos) [reply .github/workflows/test.yml:47]: should remove sorry, it was failing so was just for deubbing.

File: core/tracker/tracker.go (hunk 3 above)
Comment (corverroos) [reply core/tracker/tracker.go:70]: only explicitly name parameters if it isn't obvious from the types what they mean. In this case `bool, string, string` doesn't indicate what they mean.

File: core/tracker/tracker.go (hunk 4 above)
Comment (corverroos) [reply core/tracker/tracker.go:153]: I would argue instrumentation is a umbrella term term that includes metrics and logs and tracing.

File: app/app.go (hunk 2 above)
Comment (corverroos) [reply app/app.go:325]: in general yes, but this is such low signal code, I put it at the start with all the "prep" variables.

**Pull request merged**